### ISSUE TITLE
add bookmark jumping feature

### ISSUE THREAD

**Issue by dracodoc** [koreader/kindlepdfviewer#438]
This is a feature @houqp wrote according to my request. Because it replaced some keyboard short cut in use, I used to think it can exist as a personal hack instead of being merged into master branch. Today I reviewed the keyboard short cut and realized it can coexist with others with minor changes.

So I suggest this feature to be merged into master branch.

The feature work as follows:
When bookmarks exists, press shift + <> will jump to previous/next bookmarked page directly. My typical usage is when I finished a book, I'll jump through each bookmark to make notes. I can move forward or backward around each bookmarked page normally to make note, then press shift + > to next bookmarked page directly.

The shift + <> will not conflict with pdf reader, so it can be added directly. With crengine, currently font resizing is using this shortcut, so I suggest to move font resize to alt + up/down, then add bookmark jumping with shfit + <>. Or you can assign bookmark jumping to alt + up/down if you don't want to change the key for font resizing.

I uploaded a zip file here
https://www.box.com/s/hyrqc2gukk0ldcm29umr
which include the patch file between the kindlepdfviewer-v2012.09-235-g4df2eab version and my modified version, and the modified file itself for your reference. Note in this version I was changing font resizing to shift up/down to avoid key confilict.

**Comment by houqp** [comment]
Sorry for holding this feature from merging into master for so long... I promise I will get it done tonight :(

In order to keep user interface consistent, I will use `alt + up/down` for bookmark jumping, is that OK?

**Comment by dracodoc** [comment]
No problem, thanks! :)

**Comment by tigran123** [comment]
closing, merged into master.

**Comment by dracodoc** [comment]
I noticed the newest build of v2012.09-258-gcd5324c.zip is using alt + up/down in crengine, but alt + K/L in pdf. Is this intended? Shouldn't we use the same key if possible?

Besides, it seemed that the alt + up/down doesn't really work. Pressing alt + up/down only move the screen a little bit like just pressing up/down.

**Comment by tigran123** [comment]
@dracodoc Alt-up/down/left/right keys do not work on Kindle, so we had to use Alt-H/J (for TOC cycling) and Alt-K/L (for bookmarks cycling).

**Comment by tigran123** [comment]
@dracodoc Oh, also, on crengine I think I've deleted the Alt-K/L group instead of deleting Alt-H/J :)

Let me double-check and if I really did that then I'll fix it quickly.

**Comment by dracodoc** [comment]
In the build 09-258, crereader.lua is still using alt + up/down.

**Comment by tigran123** [comment]
Ok, done, see PR #451.

**Comment by tigran123** [comment]
@dracodoc Btw, I only tested bookmarks cycling feature with PDF files, not crengine (it wasn't me who wrote it --- it was @houqp so I assumed he tested it with crengine). When trying to use it with crengine I get crashes... I will have a look now.

**Comment by tigran123** [comment]
Ah, I see, it is my fault --- I changed the keys to Alt-K/L for PDF/DjVu but forgot to do the same for crengine.
I will fix it now.

**Comment by tigran123** [comment]
Ok, the commit with the fix added to the same PR #451

I'll throw away the build 258 and put a new build now.

**Comment by tigran123** [comment]
@dracodoc Ok, try the build 264 now --- all the fixes are there:

https://github.com/tigran123/kindlepdfviewer/downloads

**Comment by dracodoc** [comment]
It works now. 2 of other things I found:
1. "I" change the way of inform of events, but pressing "I" only get a pop up said it's unstable, then what? I didn't see any change or in which way of inform of event I'm using.
2. The font zoom operation dialog stayed about full 4 seconds before refresh, it was too long. I checked the code, it is 
InfoMessage:inform(change.." font size to "..self.font_zoom..". ", nil, 1, MSG_AUX)

I'm not sure it was suppose to be 1 second here, or my previous attempt on "I" changed this behavior?

**Comment by tigran123** [comment]
1. You have to raise the privilege level as changing the info methods requires top-level clearance. To give yourself this high privilege level you have to use `Alt-M` command (it works both within reader and file manager)
2. There is no concept of "zoom" for ebooks. When you increase or decrease the font size the engine (in this case crengine) has to retypeset (i.e. collect text into lines, lines into paragraphs, paragraphs into pages) the entire book with the new font size (or new interlinear skip or bolder font, etc). For long books it can take minutes, so 4 seconds is actually quite good (i.e. your book was small enough). This has nothing to do with any messages or timeouts.

**Comment by dracodoc** [comment]
OK, got it. I suggest to change the message a little bit for "I', because the message didn't hint that it is limited by the privilege level, instead user would only know that the function is unstable then don't know how to try it even he would like to take some risk.

**Comment by tigran123** [comment]
Yes, I was going to change it when I changed the "set file manager mode" to "set user privilege level" (because it was not just limited to file manager's operations). But I couldn't think of any better alternative. Can you suggest one?

**Comment by dracodoc** [comment]
I noticed the code is like this
InfoMessage:inform("Unstable... For experts only ", -1, 1, MSG_WARN,
                    "This function is still under development and available only for experts and beta testers.")

while when the privilege level is set for beginner, I only saw the first message "Unstable... For experts only". The second message seemed never appear even in the expert level so it seemed useless.

As for the message, how about this
"Change user privilege to expert level first to enable this experimental feature"

It's rather long but difficult to make it shorter.
